File: backend/app/schemas/pdv_terminal.py
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator


class PdvTerminalRegister(BaseModel):
    cash_register_number: str = Field(min_length=1, max_length=10)
    terminal_key: str = Field(min_length=8, max_length=180)
    app_version: str | None = Field(default=None, max_length=40)
    device_label: str | None = Field(default=None, max_length=120)

    @field_validator("cash_register_number", "terminal_key", mode="before")
    @classmethod
    def clean_required_text(cls, value: object) -> str:
        return str(value or "").strip()

    @field_validator("cash_register_number")
    @classmethod
    def normalize_cash_register_number(cls, value: str) -> str:
        digits = "".join(char for char in value if char.isdigit())
        if not digits:
            raise ValueError("Informe o numero do caixa.")
        return digits.zfill(2)[-10:]


class PdvTerminalHeartbeat(BaseModel):
    terminal_key: str = Field(min_length=8, max_length=180)
    app_version: str | None = Field(default=None, max_length=40)
    device_label: str | None = Field(default=None, max_length=120)
    current_status: str = Field(default="closed", max_length=30)
    current_operator_name: str | None = Field(default=None, max_length=150)
    cash_opened_at: datetime | None = None
    current_session_total_amount: Decimal | None = None

    @field_validator("terminal_key", "current_status", mode="before")
    @classmethod
    def clean_text(cls, value: object) -> str:
        return str(value or "").strip()

    @field_validator("current_status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        status = value.strip().lower()
        if status not in {"open", "paused", "closed"}:
            raise ValueError("Status do PDV invalido.")
        return status


class PdvTerminalNumberUpdate(BaseModel):
    cash_register_number: str = Field(min_length=1, max_length=10)

    @field_validator("cash_register_number", mode="before")
    @classmethod
    def clean_number(cls, value: object) -> str:
        return str(value or "").strip()

    @field_validator("cash_register_number")
    @classmethod
    def normalize_cash_register_number(cls, value: str) -> str:
        digits = "".join(char for char in value if char.isdigit())
        if not digits:
            raise ValueError("Informe o numero do caixa.")
        return digits.zfill(2)[-10:]


class PdvTerminalActivationCreate(BaseModel):
    cash_register_number: str = Field(min_length=1, max_length=10)
    device_label: str | None = Field(default=None, max_length=120)
    expires_hours: int = Field(default=24, ge=1, le=168)

    @field_validator("cash_register_number", mode="before")
    @classmethod
    def clean_number(cls, value: object) -> str:
        return str(value or "").strip()

    @field_validator("cash_register_number")
    @classmethod
    def normalize_cash_register_number(cls, value: str) -> str:
        digits = "".join(char for char in value if char.isdigit())
        if not digits:
            raise ValueError("Informe o numero do caixa.")
        return digits.zfill(2)[-10:]

    @field_validator("device_label", mode="before")
    @classmethod
    def clean_optional_text(cls, value: object) -> str | None:
        text = str(value or "").strip()
        return text or None


class MasterPdvTerminalActivationCreate(PdvTerminalActivationCreate):
    company_code: str = Field(min_length=1, max_length=80)

    @field_validator("company_code", mode="before")
    @classmethod
    def clean_company_code(cls, value: object) -> str:
        return str(value or "").strip().lower()


class PdvTerminalActivationCodeRead(BaseModel):
    terminal: "PdvTerminalRead"
    activation_code: str
    expires_at: datetime


class PdvTerminalActivationRequest(BaseModel):
    activation_code: str = Field(min_length=6, max_length=40)
    app_version: str | None = Field(default=None, max_length=40)
    device_label: str | None = Field(default=None, max_length=120)
    machine_name: str | None = Field(default=None, max_length=120)
    windows_user: str | None = Field(default=None, max_length=120)
    windows_version: str | None = Field(default=None, max_length=120)
    device_fingerprint: str | None = Field(default=None, max_length=180)

    @field_validator(
        "activation_code",
        "app_version",
        "device_label",
        "machine_name",
        "windows_user",
        "windows_version",
        "device_fingerprint",
        mode="before",
    )
    @classmethod
    def clean_text(cls, value: object) -> str | None:
        text = str(value or "").strip()
        return text or None


class PdvTerminalRead(BaseModel):
    id: int
    cash_register_number: str
    terminal_key: str
    app_version: str | None
    device_label: str | None
    activation_status: str = "active"
    activation_code_expires_at: datetime | None = None
    activated_at: datetime | None = None
    machine_name: str | None = None
    windows_user: str | None = None
    windows_version: str | None = None
    device_fingerprint: str | None = None
    active: bool
    current_status: str | None
    current_operator_name: str | None
    cash_opened_at: datetime | None
    current_session_total_amount: Decimal | None
    today_sales_count: int = 0
    today_sales_amount: Decimal = Decimal("0")
    created_at: datetime
    updated_at: datetime | None
    last_seen_at: datetime | None
    crossed_business_day: bool = False

    model_config = ConfigDict(from_attributes=True)


class PdvBusinessDaySettings(BaseModel):
    cutoff_minutes: int = Field(default=180, ge=0, le=1439)


class PdvTerminalCommandCreate(BaseModel):
    action: str = Field(min_length=1, max_length=40)
    message: str | None = Field(default=None, max_length=240)
    payload: dict[str, Any] | None = None

    @field_validator("action", mode="before")
    @classmethod
    def clean_action(cls, value: object) -> str:
        return str(value or "").strip().lower()

    @field_validator("action")
    @classmethod
    def validate_action(cls, value: str) -> str:
        if value not in {
            "force_reconnect",
            "sync_offline_sales",
            "clear_local_queue",
            "block_terminal",
            "unblock_terminal",
            "reset_terminal_link",
            "request_update",
            "reload_catalog",
        }:
            raise ValueError("Comando do terminal invalido.")
        return value

    @field_validator("message", mode="before")
    @classmethod
    def clean_optional_message(cls, value: object) -> str | None:
        text = str(value or "").strip()
        return text or None


class PdvTerminalCommandAck(BaseModel):
    terminal_key: str = Field(min_length=8, max_length=180)
    status: str = Field(min_length=1, max_length=20)
    result_message: str | None = Field(default=None, max_length=2000)

    @field_validator("terminal_key", "status", mode="before")
    @classmethod
    def clean_required_text(cls, value: object) -> str:
        return str(value or "").strip()

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        status = value.strip().lower()
        if status not in {"done", "failed"}:
            raise ValueError("Status do comando invalido.")
        return status

    @field_validator("result_message", mode="before")
    @classmethod
    def clean_result_message(cls, value: object) -> str | None:
        text = str(value or "").strip()
        return text or None


class PdvTerminalCommandRead(BaseModel):
    id: int
    terminal_id: int
    action: str
    status: str
    message: str | None = None
    payload: dict[str, Any] | None = None
    result_message: str | None = None
    created_at: datetime
    delivered_at: datetime | None = None
    completed_at: datetime | None = None

File: admin_app/admin_flutter/lib/models/pdv_terminal.dart
class PdvTerminal {
  const PdvTerminal({
    required this.id,
    required this.cashRegisterNumber,
    required this.terminalKey,
    required this.active,
    this.appVersion,
    this.deviceLabel,
    this.activationStatus = 'active',
    this.activationCodeExpiresAt,
    this.activatedAt,
    this.machineName,
    this.windowsUser,
    this.windowsVersion,
    this.deviceFingerprint,
    this.currentStatus,
    this.currentOperatorName,
    this.cashOpenedAt,
    this.currentSessionTotalAmount,
    this.todaySalesCount = 0,
    this.todaySalesAmount = 0,
    this.lastSeenAt,
    this.crossedBusinessDay = false,
  });

  final int id;
  final String cashRegisterNumber;
  final String terminalKey;
  final String? appVersion;
  final String? deviceLabel;
  final String activationStatus;
  final DateTime? activationCodeExpiresAt;
  final DateTime? activatedAt;
  final String? machineName;
  final String? windowsUser;
  final String? windowsVersion;
  final String? deviceFingerprint;
  final bool active;
  final String? currentStatus;
  final String? currentOperatorName;
  final DateTime? cashOpenedAt;
  final double? currentSessionTotalAmount;
  final int todaySalesCount;
  final double todaySalesAmount;
  final DateTime? lastSeenAt;
  final bool crossedBusinessDay;

  factory PdvTerminal.fromJson(Map<String, dynamic> json) {
    return PdvTerminal(
      id: json['id'] as int,
      cashRegisterNumber: json['cash_register_number'] as String,
      terminalKey: json['terminal_key'] as String,
      appVersion: json['app_version'] as String?,
      deviceLabel: json['device_label'] as String?,
      activationStatus: json['activation_status'] as String? ?? 'active',
      activationCodeExpiresAt: json['activation_code_expires_at'] == null
          ? null
          : DateTime.parse(
              json['activation_code_expires_at'] as String,
            ).toLocal(),
      activatedAt: json['activated_at'] == null
          ? null
          : DateTime.parse(json['activated_at'] as String).toLocal(),
      machineName: json['machine_name'] as String?,
      windowsUser: json['windows_user'] as String?,
      windowsVersion: json['windows_version'] as String?,
      deviceFingerprint: json['device_fingerprint'] as String?,
      active: json['active'] as bool? ?? true,
      currentStatus: json['current_status'] as String?,
      currentOperatorName: json['current_operator_name'] as String?,
      cashOpenedAt: json['cash_opened_at'] == null
          ? null
          : DateTime.parse(json['cash_opened_at'] as String).toLocal(),
      currentSessionTotalAmount: _toDouble(
        json['current_session_total_amount'],
      ),
      todaySalesCount: json['today_sales_count'] as int? ?? 0,
      todaySalesAmount: _toDouble(json['today_sales_amount']) ?? 0,
      lastSeenAt: json['last_seen_at'] == null
          ? null
          : DateTime.parse(json['last_seen_at'] as String).toLocal(),
      crossedBusinessDay: json['crossed_business_day'] as bool? ?? false,
    );
  }

  static double? _toDouble(Object? value) {
    if (value == null) return null;
    if (value is num) return value.toDouble();
    return double.tryParse(value.toString());
  }
}

class CompanyResourceQuota {
  const CompanyResourceQuota({
    required this.companyCode,
    required this.activeUsers,
    this.maxUsers,
    required this.pdvTerminals,
    this.maxPdvTerminals,
  });

  final String companyCode;
  final int activeUsers;
  final int? maxUsers;
  final int pdvTerminals;
  final int? maxPdvTerminals;

  bool get pdvLimitReached =>
      maxPdvTerminals != null && pdvTerminals >= maxPdvTerminals!;

  factory CompanyResourceQuota.fromJson(Map<String, dynamic> json) {
    return CompanyResourceQuota(
      companyCode: json['company_code'] as String,
      activeUsers: json['active_users'] as int? ?? 0,
      maxUsers: json['max_users'] as int?,
      pdvTerminals: json['pdv_terminals'] as int? ?? 0,
      maxPdvTerminals: json['max_pdv_terminals'] as int?,
    );
  }
}

class PdvBusinessDaySettings {
  const PdvBusinessDaySettings({required this.cutoffMinutes});

  final int cutoffMinutes;

  factory PdvBusinessDaySettings.fromJson(Map<String, dynamic> json) {
    return PdvBusinessDaySettings(
      cutoffMinutes: json['cutoff_minutes'] as int? ?? 180,
    );
  }
}

class PdvTerminalActivationCreatePayload {
  const PdvTerminalActivationCreatePayload({
    required this.cashRegisterNumber,
    this.deviceLabel,
    this.expiresHours = 24,
  });

  final String cashRegisterNumber;
  final String? deviceLabel;
  final int expiresHours;

  Map<String, dynamic> toJson() {
    return {
      'cash_register_number': cashRegisterNumber,
      'device_label': deviceLabel,
      'expires_hours': expiresHours,
    };
  }
}

class MasterPdvTerminalActivationCreatePayload
    extends PdvTerminalActivationCreatePayload {
  const MasterPdvTerminalActivationCreatePayload({
    required this.companyCode,
    required super.cashRegisterNumber,
    super.deviceLabel,
    super.expiresHours,
  });

  final String companyCode;

  @override
  Map<String, dynamic> toJson() {
    return {...super.toJson(), 'company_code': companyCode};
  }
}

class PdvTerminalActivationCode {
  const PdvTerminalActivationCode({
    required this.terminal,
    required this.activationCode,
    required this.expiresAt,
  });

  final PdvTerminal terminal;
  final String activationCode;
  final DateTime expiresAt;

  factory PdvTerminalActivationCode.fromJson(Map<String, dynamic> json) {
    return PdvTerminalActivationCode(
      terminal: PdvTerminal.fromJson(json['terminal'] as Map<String, dynamic>),
      activationCode: json['activation_code'] as String,
      expiresAt: DateTime.parse(json['expires_at'] as String).toLocal(),
    );
  }
}

class PdvTerminalRegisterPayload {
  const PdvTerminalRegisterPayload({
    required this.cashRegisterNumber,
    required this.terminalKey,
    this.appVersion,
    this.deviceLabel,
  });

  final String cashRegisterNumber;
  final String terminalKey;
  final String? appVersion;
  final String? deviceLabel;

  Map<String, dynamic> toJson() {
    return {
      'cash_register_number': cashRegisterNumber,
      'terminal_key': terminalKey,
      'app_version': appVersion,
      'device_label': deviceLabel,
    };
  }
}

class PdvTerminalHeartbeatPayload {
  const PdvTerminalHeartbeatPayload({
    required this.terminalKey,
    required this.currentStatus,
    this.appVersion,
    this.deviceLabel,
    this.currentOperatorName,
    this.cashOpenedAt,
    this.currentSessionTotalAmount,
  });

  final String terminalKey;
  final String currentStatus;
  final String? appVersion;
  final String? deviceLabel;
  final String? currentOperatorName;
  final DateTime? cashOpenedAt;
  final double? currentSessionTotalAmount;

  Map<String, dynamic> toJson() {
    return {
      'terminal_key': terminalKey,
      'app_version': appVersion,
      'device_label': deviceLabel,
      'current_status': currentStatus,
      'current_operator_name': currentOperatorName,
      'cash_opened_at': cashOpenedAt?.toIso8601String(),
      'current_session_total_amount': currentSessionTotalAmount,
    };
  }
}

class PdvTerminalCommand {
  const PdvTerminalCommand({
    required this.id,
    required this.terminalId,
    required this.action,
    required this.status,
    this.message,
    this.resultMessage,
    this.createdAt,
    this.deliveredAt,
    this.completedAt,
  });

  final int id;
  final int terminalId;
  final String action;
  final String status;
  final String? message;
  final String? resultMessage;
  final DateTime? createdAt;
  final DateTime? deliveredAt;
  final DateTime? completedAt;

  factory PdvTerminalCommand.fromJson(Map<String, dynamic> json) {
    return PdvTerminalCommand(
      id: json['id'] as int,
      terminalId: json['terminal_id'] as int,
      action: json['action'] as String,
      status: json['status'] as String,
      message: json['message'] as String?,
      resultMessage: json['result_message'] as String?,
      createdAt: json['created_at'] == null
          ? null
          : DateTime.parse(json['created_at'] as String).toLocal(),
      deliveredAt: json['delivered_at'] == null
          ? null
          : DateTime.parse(json['delivered_at'] as String).toLocal(),
      completedAt: json['completed_at'] == null
          ? null
          : DateTime.parse(json['completed_at'] as String).toLocal(),
    );
  }
}
